src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,172 @@
+   *  Visible       Hidden        
+   *  Layer         Layer         
+   * +-----+       +-----+       
+   * |     |       |     |            
+   * |     +<----->|     |
+   * |     |       |     | 
+   * +-----+       +-----+     
+   *        
+   */
+  arma::mat output;
+  BinaryLayer<> visible(trainData.n_rows, hiddenLayerSize, 1);
+  BinaryLayer<> hidden(hiddenLayerSize, trainData.n_rows, 0);
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, BinaryLayer<>, BinaryLayer<>> model(trainData,
+      gaussian, visible, hidden, 1,  true);
+  MiniBatchSGD msgd(10, 0.06, trainData.n_cols * 20, 0, true);
+  model.Train(trainData, msgd);
+  model.Reset();
+  // Set the parmaeters from a learned rbm
+  model.Parameters() = arma::mat(
+      "-0.23224054, -0.23000632, -0.25701271, -0.25122418, -0.20716651"
+      "-0.20962217, -0.59922456, -0.60003836, -0.6, -0.625, -0.475;");
+  // Check Weight Shared
+  BOOST_REQUIRE_CLOSE(arma::accu(model.VisibleLayer().Weight() -
+      model.HiddenLayer().Weight()), 0, 1e-2);
+
+  // Check free energy
+  arma::vec freeEnergy = arma::mat(
+      "-0.87523715, 0.50615066, 0.46923476, 1.21509084;");
+  arma::vec calcultedFreeEnergy(4);
+  calcultedFreeEnergy.zeros();
+  for (size_t i = 0; i < trainData.n_cols; i++)
+  {
+    model.VisibleLayer().ForwardPreActivation(std::move(trainData.col(i)),
+        std::move(output));
+    calcultedFreeEnergy(i) = model.FreeEnergy(std::move(trainData.col(i)));
+  }
+
+  BOOST_REQUIRE_LE(arma::accu(calcultedFreeEnergy - freeEnergy), 1);
+}
+BOOST_AUTO_TEST_CASE(MiscTest)
+{
+  /**
+   * Train and evaluate a vanilla network with the specified structure.
+   */
+
+  arma::mat X = arma::mat("0, 0, 0;"
+                          "0, 1, 1;"
+                          "1, 0, 1;"
+                          "1, 1, 1;");
+  X = X.t();
+  BuildVanillaNetwork<>(X, 2);
+}
+
+BOOST_AUTO_TEST_CASE(ClassificationTest)
+{
+  // Normalised dataset
+  int hiddenLayerSize = 100;
+  arma::mat trainData, testData, dataset;
+  arma::vec trainLabelsTemp, testLabelsTemp;
+  data::Load("mnist_data_small.csv", dataset, true);
+  trainData = dataset.submat(0, 0,
+      dataset.n_rows - 102, dataset.n_cols - 2);
+  testData = dataset.submat(dataset.n_rows - 102, 0 ,
+      dataset.n_rows - 2, dataset.n_cols - 2);
+  trainLabelsTemp = dataset.submat(0, dataset.n_cols - 1,
+      dataset.n_rows - 102, dataset.n_cols - 1);
+  testLabelsTemp = dataset.submat(dataset.n_rows - 102, dataset.n_cols - 1,
+      dataset.n_rows - 2, dataset.n_cols - 1);
+
+  arma::Row<size_t> trainLabels = arma::zeros<arma::Row<size_t>>(1,
+      trainLabelsTemp.n_rows);
+  arma::Row<size_t> testLabels = arma::zeros<arma::Row<size_t>>(1,
+      testLabelsTemp.n_rows);
+
+  for (size_t i = 0; i < trainLabelsTemp.n_rows; ++i)
+  {
+    trainLabels(i) = arma::as_scalar(trainLabelsTemp.row(i));
+  }
+
+  for (size_t i = 0; i < testLabels.n_rows; ++i)
+  {
+    testLabels(i) = arma::as_scalar(testLabels.row(i));
```

Comment:
You don't initialize `testLabels` properly. You initialize `testLabelsTemp` but you don't copy that to `testLabels`, so `testLabels` is equal to zero. Moreover `arma::as_scalar(testLabels.row(i))` is equal to `testLabels(i)` since `testLabels` is a vector.